### PR TITLE
Convert limit specific SQL language for Oracle 

### PR DESCRIPTION
**Pull request by louis878**
**Is your feature request related to a problem? Please describe.**
We build queries using the generic 'limit X offset Y' syntax. Some database technologies do not use this syntax. The generic syntax should be refactored to the DB specific language where needed.

**Id**
125f575

### CODE REVIEW

File: core/java/iesi-core/src/main/java/io/metadew/iesi/connection/database/db2/Db2DatabaseConnectionService.java [hunk 1]
```diff
@@ -28,4 +29,8 @@ public String getDriver(Db2DatabaseConnection databaseConnection) {
         return Db2DatabaseConnection.class;
     }
 
+    @Override
+    public String refactorLimitAndOffset(Db2DatabaseConnection databaseConnection, String query) {
+        return query;
+    }
```
Comment (BerrevoetsRobbe) [inline core/java/iesi-core/src/main/java/io/metadew/iesi/connection/database/db2/Db2DatabaseConnectionService.java:35]: All XDatabaseConnectionService's extend either DatabaseConnectionService or SchemaDatabaseConnectionService (which extends DatabaseConnectionService). You could implement the default behavior in the DatabaseConnectionService  class

File: core/java/iesi-core/src/main/java/io/metadew/iesi/connection/database/dremio/DremioDatabaseConnectionService.java [hunk 2]
```diff
@@ -27,4 +27,9 @@ public String getDriver(DremioDatabaseConnection databaseConnection) {
     public Class<DremioDatabaseConnection> appliesTo() {
         return DremioDatabaseConnection.class;
     }
+
+    @Override
+    public String refactorLimitAndOffset(DremioDatabaseConnection databaseConnection, String query) {
+        return query;
+    }
```
Comment (BerrevoetsRobbe) [inline core/java/iesi-core/src/main/java/io/metadew/iesi/connection/database/dremio/DremioDatabaseConnectionService.java:34]: All XDatabaseConnectionService's extend either DatabaseConnectionService or SchemaDatabaseConnectionService (which extends DatabaseConnectionService). You could implement the default behavior in the DatabaseConnectionService class

File: core/java/iesi-core/src/main/java/io/metadew/iesi/connection/database/drill/DrillDatabaseConnectionService.java [hunk 3]
```diff
@@ -27,4 +27,9 @@ public String getDriver(DrillDatabaseConnection databaseConnection) {
     public Class<DrillDatabaseConnection> appliesTo() {
         return DrillDatabaseConnection.class;
     }
+
+    @Override
+    public String refactorLimitAndOffset(DrillDatabaseConnection databaseConnection, String query) {
+        return query;
+    }
```
Comment (BerrevoetsRobbe) [inline core/java/iesi-core/src/main/java/io/metadew/iesi/connection/database/drill/DrillDatabaseConnectionService.java:34]: All XDatabaseConnectionService's extend either DatabaseConnectionService or SchemaDatabaseConnectionService (which extends DatabaseConnectionService). You could implement the default behavior in the DatabaseConnectionService class

File: core/java/iesi-core/src/main/java/io/metadew/iesi/connection/database/h2/H2DatabaseConnectionService.java [hunk 4]
```diff
@@ -32,6 +32,11 @@ public String getDriver(H2DatabaseConnection databaseConnection) {
         return H2DatabaseConnection.class;
     }
 
+    @Override
+    public String refactorLimitAndOffset(H2DatabaseConnection databaseConnection, String query) {
+        return query;
+    }
+
```
Comment (BerrevoetsRobbe) [inline core/java/iesi-core/src/main/java/io/metadew/iesi/connection/database/h2/H2DatabaseConnectionService.java:39]: All XDatabaseConnectionService's extend either DatabaseConnectionService or SchemaDatabaseConnectionService (which extends DatabaseConnectionService). You could implement the default behavior in the DatabaseConnectionService class

File: core/java/iesi-core/src/main/java/io/metadew/iesi/connection/database/mariadb/MariadbDatabaseConnectionService.java [hunk 5]
```diff
@@ -28,4 +30,8 @@ public String getDriver(MariadbDatabaseConnection databaseConnection) {
         return MariadbDatabaseConnection.class;
     }
 
+    @Override
+    public String refactorLimitAndOffset(MariadbDatabaseConnection databaseConnection, String query) {
+        return query;
+    }
```
Comment (BerrevoetsRobbe) [inline core/java/iesi-core/src/main/java/io/metadew/iesi/connection/database/mariadb/MariadbDatabaseConnectionService.java:36]: All XDatabaseConnectionService's extend either DatabaseConnectionService or SchemaDatabaseConnectionService (which extends DatabaseConnectionService). You could implement the default behavior in the DatabaseConnectionService class

File: core/java/iesi-core/src/main/java/io/metadew/iesi/connection/database/mssql/MssqlDatabaseConnectionService.java [hunk 6]
```diff
@@ -27,4 +27,9 @@ public String getDriver(MssqlDatabaseConnection databaseConnection) {
     public Class<MssqlDatabaseConnection> appliesTo() {
         return MssqlDatabaseConnection.class;
     }
+
+    @Override
+    public String refactorLimitAndOffset(MssqlDatabaseConnection databaseConnection, String query) {
+        return query;
+    }
```
Comment (BerrevoetsRobbe) [inline core/java/iesi-core/src/main/java/io/metadew/iesi/connection/database/mssql/MssqlDatabaseConnectionService.java:34]: All XDatabaseConnectionService's extend either DatabaseConnectionService or SchemaDatabaseConnectionService (which extends DatabaseConnectionService). You could implement the default behavior in the DatabaseConnectionService class

File: core/java/iesi-core/src/main/java/io/metadew/iesi/connection/database/mysql/MysqlDatabaseConnectionService.java [hunk 7]
```diff
@@ -27,4 +27,9 @@ public String getDriver(MysqlDatabaseConnection databaseConnection) {
     public Class<MysqlDatabaseConnection> appliesTo() {
         return MysqlDatabaseConnection.class;
     }
+
+    @Override
+    public String refactorLimitAndOffset(MysqlDatabaseConnection databaseConnection, String query) {
+        return query;
+    }
```
Comment (BerrevoetsRobbe) [inline core/java/iesi-core/src/main/java/io/metadew/iesi/connection/database/mysql/MysqlDatabaseConnectionService.java:34]: All XDatabaseConnectionService's extend either DatabaseConnectionService or SchemaDatabaseConnectionService (which extends DatabaseConnectionService). You could implement the default behavior in the DatabaseConnectionService class

File: core/java/iesi-core/src/main/java/io/metadew/iesi/connection/database/netezza/NetezzaDatabaseConnectionService.java [hunk 8]
```diff
@@ -27,4 +27,9 @@ public String getDriver(NetezzaDatabaseConnection databaseConnection) {
     public Class<NetezzaDatabaseConnection> appliesTo() {
         return NetezzaDatabaseConnection.class;
     }
+
+    @Override
+    public String refactorLimitAndOffset(NetezzaDatabaseConnection databaseConnection, String query) {
+        return query;
+    }
```
Comment (BerrevoetsRobbe) [inline core/java/iesi-core/src/main/java/io/metadew/iesi/connection/database/netezza/NetezzaDatabaseConnectionService.java:34]: All XDatabaseConnectionService's extend either DatabaseConnectionService or SchemaDatabaseConnectionService (which extends DatabaseConnectionService). You could implement the default behavior in the DatabaseConnectionService class

File: core/java/iesi-core/src/main/java/io/metadew/iesi/connection/database/oracle/OracleDatabaseConnectionService.java [hunk 9]
```diff
@@ -56,4 +56,9 @@ public Connection getConnection(OracleDatabaseConnection oracleDatabaseConnectio
             throw new RuntimeException(e);
         }
     }
+
+    public String refactorLimitAndOffset(OracleDatabaseConnection databaseConnection, String query) {
+        String newQuery = query.replaceAll("(limit|LIMIT|Limit)", "OffSet").replaceAll("(offset|OFFSET|Offset)", "ROWS FETCH NEXT").replaceAll("(ROWS FETCH NEXT\\s+\\w+)", "  $1 ROWS ONLY ");
```
Comment (BerrevoetsRobbe) [inline core/java/iesi-core/src/main/java/io/metadew/iesi/connection/database/oracle/OracleDatabaseConnectionService.java:61]: Is it not possible to write a regex for the entire 'limit X offset Y' clause and replace it with a 'offset Y rows fetch next X rows only'? Otherwise the simple occurrence of limit or offset will be replaced

File: core/java/iesi-core/src/main/java/io/metadew/iesi/connection/database/postgresql/PostgresqlDatabaseConnectionService.java [hunk 10]
```diff
@@ -27,4 +27,9 @@ public String getDriver(PostgresqlDatabaseConnection databaseConnection) {
     public Class<PostgresqlDatabaseConnection> appliesTo() {
         return PostgresqlDatabaseConnection.class;
     }
+
+    @Override
+    public String refactorLimitAndOffset(PostgresqlDatabaseConnection databaseConnection, String query) {
+        return query;
+    }
```
Comment (BerrevoetsRobbe) [inline core/java/iesi-core/src/main/java/io/metadew/iesi/connection/database/postgresql/PostgresqlDatabaseConnectionService.java:34]: All XDatabaseConnectionService's extend either DatabaseConnectionService or SchemaDatabaseConnectionService (which extends DatabaseConnectionService). You could implement the default behavior in the DatabaseConnectionService class

File: core/java/iesi-core/src/main/java/io/metadew/iesi/connection/database/presto/PrestoDatabaseConnectionService.java [hunk 11]
```diff
@@ -27,4 +27,9 @@ public String getDriver(PrestoDatabaseConnection databaseConnection) {
     public Class<PrestoDatabaseConnection> appliesTo() {
         return PrestoDatabaseConnection.class;
     }
+
+    @Override
+    public String refactorLimitAndOffset(PrestoDatabaseConnection databaseConnection, String query) {
+        return query;
+    }
```
Comment (BerrevoetsRobbe) [inline core/java/iesi-core/src/main/java/io/metadew/iesi/connection/database/presto/PrestoDatabaseConnectionService.java:34]: All XDatabaseConnectionService's extend either DatabaseConnectionService or SchemaDatabaseConnectionService (which extends DatabaseConnectionService). You could implement the default behavior in the DatabaseConnectionService class

File: core/java/iesi-core/src/main/java/io/metadew/iesi/connection/database/sqlite/SqliteDatabaseConnectionService.java [hunk 12]
```diff
@@ -34,4 +36,9 @@ public String getDriver(SqliteDatabaseConnection databaseConnection) {
     public Class<SqliteDatabaseConnection> appliesTo() {
         return SqliteDatabaseConnection.class;
     }
+
+    @Override
+    public String refactorLimitAndOffset(SqliteDatabaseConnection databaseConnection, String query) {
+        return query;
+    }
```
Comment (BerrevoetsRobbe) [inline core/java/iesi-core/src/main/java/io/metadew/iesi/connection/database/sqlite/SqliteDatabaseConnectionService.java:43]: All XDatabaseConnectionService's extend either DatabaseConnectionService or SchemaDatabaseConnectionService (which extends DatabaseConnectionService). You could implement the default behavior in the DatabaseConnectionService class

File: core/java/iesi-core/src/main/java/io/metadew/iesi/connection/database/temporary/TemporaryDatabaseConnectionServiceImpl.java [hunk 13]
```diff
@@ -127,4 +127,9 @@ public PreparedStatement createPreparedStatement(TemporaryDatabaseConnection dat
     public Class<TemporaryDatabaseConnection> appliesTo() {
         return TemporaryDatabaseConnection.class;
     }
+
+    @Override
+    public String refactorLimitAndOffset(TemporaryDatabaseConnection databaseConnection, String query) {
+        return query;
+    }
```
Comment (BerrevoetsRobbe) [inline core/java/iesi-core/src/main/java/io/metadew/iesi/connection/database/temporary/TemporaryDatabaseConnectionServiceImpl.java:134]: All XDatabaseConnectionService's extend either DatabaseConnectionService or SchemaDatabaseConnectionService (which extends DatabaseConnectionService). You could implement the default behavior in the DatabaseConnectionService class

File: core/java/iesi-core/src/main/java/io/metadew/iesi/connection/database/teradata/TeradataDatabaseConnectionService.java [hunk 14]
```diff
@@ -27,4 +27,9 @@ public String getDriver(TeradataDatabaseConnection databaseConnection) {
     public Class<TeradataDatabaseConnection> appliesTo() {
         return TeradataDatabaseConnection.class;
     }
+
+    @Override
+    public String refactorLimitAndOffset(TeradataDatabaseConnection databaseConnection, String query) {
+        return query;
+    }
```
Comment (BerrevoetsRobbe) [inline core/java/iesi-core/src/main/java/io/metadew/iesi/connection/database/teradata/TeradataDatabaseConnectionService.java:34]: All XDatabaseConnectionService's extend either DatabaseConnectionService or SchemaDatabaseConnectionService (which extends DatabaseConnectionService). You could implement the default behavior in the DatabaseConnectionService class

File: core/java/iesi-rest-without-microservices/src/test/resources/application-iesi-repository.yml [hunk 15]
```diff
@@ -4,13 +4,13 @@ iesi:
     repository:
       # category can be: general, design, connectivity, trace, result.
       # each category should define its own type and definition
-      - categories:
+        categories:
```
Comment (BerrevoetsRobbe) [inline core/java/iesi-rest-without-microservices/src/test/resources/application-iesi-repository.yml:7]: why is this removed?

File: core/java/iesi-rest-without-microservices/src/test/resources/application-iesi-repository.yml [hunk 16]
```diff
@@ -4,13 +4,13 @@ iesi:
...
           - general
         instance:
 
         # type of database: sqlite, h2, mssql, netezza, oracle, postgresql
         coordinator:
-          type: h2
+          type: oracle
```
Comment (BerrevoetsRobbe) [inline core/java/iesi-rest-without-microservices/src/test/resources/application-iesi-repository.yml:13]: why is this changed?

File: core/java/iesi-core/src/main/java/io/metadew/iesi/connection/database/oracle/OracleDatabaseConnectionService.java (hunk 9 above)
Comment (louis878) [reply core/java/iesi-core/src/main/java/io/metadew/iesi/connection/database/oracle/OracleDatabaseConnectionService.java]: Done

File: core/java/iesi-core/src/main/java/io/metadew/iesi/connection/database/oracle/OracleDatabaseConnectionService.java [hunk 17]
```diff
@@ -56,4 +58,13 @@ public Connection getConnection(OracleDatabaseConnection oracleDatabaseConnectio
             throw new RuntimeException(e);
         }
     }
+
+    public String refactorLimitAndOffset(String query) {
+        Pattern pattern = Pattern.compile("(?i)limit(?=\\s+[0-9]+\\s+(?i)offset+\\s+[0-9])");
```
Comment (BerrevoetsRobbe) [inline core/java/iesi-core/src/main/java/io/metadew/iesi/connection/database/oracle/OracleDatabaseConnectionService.java:63]: you can name the groups after 'limit' and 'offset' using (?<name>\d+). This way you can extract the limit value and offset value from the mather.

File: core/java/iesi-core/src/main/java/io/metadew/iesi/connection/database/oracle/OracleDatabaseConnectionService.java [hunk 18]
```diff
@@ -56,4 +58,13 @@ public Connection getConnection(OracleDatabaseConnection oracleDatabaseConnectio
...
+        Matcher matcher = pattern.matcher(query);
+        if (matcher.find()) {
+            return query.replaceAll("(?i)limit", "OffSet").replaceAll("(offset|OFFSET|Offset)", "ROWS FETCH NEXT").replaceAll("(ROWS FETCH NEXT\\s+\\w+)", " $1 ROWS ONLY ");
```
Comment (BerrevoetsRobbe) [inline core/java/iesi-core/src/main/java/io/metadew/iesi/connection/database/oracle/OracleDatabaseConnectionService.java:66]: you can use matcher.start() and matcher.end() to find the beginning and end of the 'limit X offset Y' clause. You can then inject the 'offset Y rows fetch next X rows only' clause using substring(). Note: you've also switched replace values:
offset -> offset rows
limit -> fetch next rows only

File: core/java/iesi-core/src/main/java/io/metadew/iesi/connection/database/oracle/OracleDatabaseConnectionService.java [hunk 19]
```diff
@@ -56,4 +58,13 @@ public Connection getConnection(OracleDatabaseConnection oracleDatabaseConnectio
             throw new RuntimeException(e);
         }
     }
+
+    public String refactorLimitAndOffset(String query) {
+        Pattern pattern = Pattern.compile("(?i)limit(?=\\s+[0-9]+\\s+(?i)offset+\\s+[0-9])");
+        Matcher matcher = pattern.matcher(query);
+        if (matcher.find()) {
```
Comment (BerrevoetsRobbe) [inline core/java/iesi-core/src/main/java/io/metadew/iesi/connection/database/oracle/OracleDatabaseConnectionService.java:65]: this should be a while loop as query can contain multiple limit/offset clauses

File: core/java/iesi-core/src/main/java/io/metadew/iesi/connection/database/oracle/OracleDatabaseConnectionService.java [hunk 20]
```diff
@@ -56,4 +58,13 @@ public Connection getConnection(OracleDatabaseConnection oracleDatabaseConnectio
             throw new RuntimeException(e);
         }
     }
+
+    public String refactorLimitAndOffset(String query) {
```
Comment (BerrevoetsRobbe) [inline core/java/iesi-core/src/main/java/io/metadew/iesi/connection/database/oracle/OracleDatabaseConnectionService.java:62]: Write unit tests for the function

**Review by BerrevoetsRobbe — COMMENTED** [review]
Approved
